### PR TITLE
feat: [CDS-53176]: Added Azure Repo Support using SCM service.

### PR DESCRIPTION
**Pull request by GoelPratyush99**
## Describe your changes
Azure Repo store was not supported with OPTIMIZED_GIT_FETCH_FILES FF. We have updated the go-scm repo from drone to support this functionality and updated its version.
Screenshot 2023-02-21 at 1 06 58 AM

## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>

You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
    - CodeFormat: `trigger codeformat`
    - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- TI-ALL: `trigger tiAll`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/44124)

### CODE REVIEW

**Comment by GoelPratyush99** [comment]
trigger messagecheck

**Review by abhinavsingh1196 — COMMENTED** [review]
let's create tests which will test all cases of nested directories (w/ and w/o files) 

> Reviewed 5 of 6 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/44124#-:-NOmdRIZAI766t8iCCSL:bd7fw8z)* status: 6 file(s) to review. 0 of 2 approvals obtained. 

___
*[`952-scm-java-client/src/main/java/io/harness/impl/scm/GetFilesInFolderForkTask.java` line 88 at r1](https://reviewable.io/reviews/harness/harness-core/44124#-NOmbw9E7m0Ms8_zbb40:-NOmbw9E7m0Ms8_zbb41:bq3k9xn) ([raw file](https://github.com/harness/harness-core/blob/c23c06f6d6ef46a02b0d3fbed21d185567ceae06/952-scm-java-client/src/main/java/io/harness/impl/scm/GetFilesInFolderForkTask.java#L88)):*
> ```Java
>         List<FileChange> fileChangeList = filesInBranchResponse.getFileList();
>         if (findFilesInBranchRequest.getProvider().hasAzure()) {
>           fileChangeList = fileChangeList.subList(1, fileChangeList.size());
> ```

Minor: Just thinking whether it will be better (and feasible?) to filter out the calling directory based on something like (item.isFolder() && item.getPath() == inputPath)
Not a big fan of filtering out the calling directory blindly, even though it may be working fine
___
*[`952-scm-java-client/src/test/java/io/harness/impl/scm/GetFilesInFolderForkTaskTest.java` line 110 at r1](https://reviewable.io/reviews/harness/harness-core/44124#-NOmcjKC2HV0c24rJm2u:-NOmcjKC2HV0c24rJm2v:bxkcnxl) ([raw file](https://github.com/harness/harness-core/blob/c23c06f6d6ef46a02b0d3fbed21d185567ceae06/952-scm-java-client/src/test/java/io/harness/impl/scm/GetFilesInFolderForkTaskTest.java#L110)):*
> ```Java
> 
>   @Test
>   @Owner(developers = ABOSII)
> ```

nit: wrong author :)

**Review by GoelPratyush99 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/44124)* status: 6 file(s) to review. 1. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`952-scm-java-client/src/main/java/io/harness/impl/scm/GetFilesInFolderForkTask.java` line 88 at r1](https://reviewable.io/reviews/harness/harness-core/44124#-NOmbw9E7m0Ms8_zbb40:-NOnV0BhCPqwEbWMTOjl:b-886gsx) ([raw file](https://github.com/harness/harness-core/blob/c23c06f6d6ef46a02b0d3fbed21d185567ceae06/952-scm-java-client/src/main/java/io/harness/impl/scm/GetFilesInFolderForkTask.java#L88)):*
<details><summary><i>Previously, abhinavsingh1196 (Abhinav Singh) wrote…</i></summary><blockquote>

Minor: Just thinking whether it will be better (and feasible?) to filter out the calling directory based on something like (item.isFolder() && item.getPath() == inputPath)
Not a big fan of filtering out the calling directory blindly, even though it may be working fine
</blockquote></details>

Fixed
___
*[`952-scm-java-client/src/test/java/io/harness/impl/scm/GetFilesInFolderForkTaskTest.java` line 110 at r1](https://reviewable.io/reviews/harness/harness-core/44124#-NOmcjKC2HV0c24rJm2u:-NOnUzj8-yynjkCptYlQ:b499q7j) ([raw file](https://github.com/harness/harness-core/blob/c23c06f6d6ef46a02b0d3fbed21d185567ceae06/952-scm-java-client/src/test/java/io/harness/impl/scm/GetFilesInFolderForkTaskTest.java#L110)):*
<details><summary><i>Previously, abhinavsingh1196 (Abhinav Singh) wrote…</i></summary><blockquote>

nit: wrong author :) 
</blockquote></details>

Changed

File: 952-scm-java-client/src/main/java/io/harness/impl/scm/GetFilesInFolderForkTask.java [hunk 1]
```diff
@@ -70,10 +70,12 @@ protected List<FileChange> compute() {
     FindFilesInBranchRequest.Builder findFilesInBranchRequest = prepareRequestBuilder(updatedFolderPath);
 
     List<FileChange> filesInBranch = getAllFilesPresentInFolder(findFilesInBranchRequest);
-    List<String> newFoldersToBeProcessed = getListOfNewFoldersToBeProcessed(filesInBranch);
-    List<GetFilesInFolderForkTask> tasksForSubFolders = createTasksForSubFolders(newFoldersToBeProcessed);
+    if (!provider.hasAzure()) {
```
Comment (abosii-harness) [inline 952-scm-java-client/src/main/java/io/harness/impl/scm/GetFilesInFolderForkTask.java:73]: Let's comment why we're doing this

**Review by abosii-harness — CHANGES_REQUESTED** [review]

**Review by abosii-harness — APPROVED** [review]
> Reviewed 4 of 6 files at r1, 1 of 2 files at r2, 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/44124)* status: 2 file(s) to review. 1. unresolved discussions. 1 of 2 approvals obtained.

**Review by abhinavsingh1196 — APPROVED** [review]
> Reviewed 1 of 2 files at r2, 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/44124)* status: :shipit: complete! ✓

**Comment by GoelPratyush99** [comment]
trigger tiAll

**Pull request merged**